git branch feature integration

tinygiant98 opened the pull request:
New branch for git integration, rebased on 0.12.3.  Incorporates requests and suggestions from #50 and addresses #27.

Instead of forcing an initial commit to recognize the repo, nasher will now do what the user is asking and provide a warning if no commits were made in the repo and the branch is not master.

If an existing branch cannot be checked out, nasher will throw a fatal error to prevent overwriting data on the current branch.  The error provided by git will be displayed to the user as part of standard nasher output.  In the sample output below, the first error is the raw git output when a checkout command fails, the second is error text from nasher.
```
      Error: error: Your local changes to the following files would be overwritten by checkout:
         ...    src/aps_onload.nss
         ... Please commit your changes or stash them before you switch branches.
         ... Aborting
      Error: You must resolve the git repo error before you can continue on branch master
```

Lots of formatting changes and new convenience functions in git.  Feel free to wail away on input and I'll get it into the format/function that you want it in.

squattingmonk commented on src/nasher/utils/git.nim:
```diff
@@ -38,13 +36,80 @@ proc gitRemote*(dir = getCurrentDir()): string =
+        else:
+          fatal("operation aborted by user")
+      else:
+        if branch.create(repo): result = branch
```
This should be `repo.create(branch)`.

tinygiant98 replied:
Yep, used to have `branch.create(repo)`, then figured `repo.create(branch)` read better, but missed one. Good catch!

squattingmonk reviewed (commented):
There's a few issues, but overall I like the approach. It looks like it will stay out of the user's way and provide good feedback.

tinygiant98 commented on src/nasher/utils/options.nim:
```diff
@@ -9,15 +9,15 @@ type
 
   Package = object
-    name*, description*, version*, url*, modName*, modMinGameVersion*: string
+    name*, description*, version*, url*, modName*, modMinGameVersion*, branch*: string
```
Ok, I think I got the `.cfg` options back into `options.nim`.  Should be good to go.

tinygiant98 commented on src/nasher/utils/git.nim:
```diff
@@ -1,12 +1,15 @@
   let (output, errCode) = execCmdEx(cmd)
   if errCode != 0:
-    default
+    if default.len == 0: output.strip
```
Restored previous function and added a carrier for the last error, if there was one, since that's what I was looking for with that change.  Not sure if the methodology is "good" programming in nim, so feel free to tell me to find another way if you don't like it.

squattingmonk commented:
I think the best route would be to use an option type. [`std/options`](https://nim-lang.org/docs/options.html) would allow something like the following:
```nim
import std/[os, osproc, options, strformat, strutils, uri]

from shared import withDir

proc gitExecCmd(cmd: string): Option[string] =
  ## Executes ``cmd`` and returns `some(output)` if successful. Otherwise, returns `none`.
  let (output, errcode) = execCmdEx(cmd)
  if errcode == 0:
    result = some(output.strip)

proc gitUser*: string =
  ## Returns the configured git username or `""` on failure.
  gitExecCmd("git config --get user.name").get("")

# ...

proc gitRemote*(repo = getCurrentDir()): string =
  ## Returns the remote for the git project in ``dir``. Supports ssh formatted
  ## remotes.
  withDir(repo):
    let url = gitExecCmd("git ls-remote --get-url")
    if url.isSome:
      result = url.get
      if result.endsWith(".git"):
        result.setLen(result.len - 4)

      if result.parseUri.scheme == "":
        let ssh = parseUri("ssh://" & result)
        result = ("https://$1/$2$3") % [ssh.hostname, ssh.port, ssh.path]

# ...

proc exists(branch: string, repo: string): bool =
  # Check for branch existence
  withDir(repo):
    gitExecCmd("git show-ref --verify refs/heads/" & branch).isSome

proc checkout(branch: string, repo: string, create = false): bool = 
  # Checkout desired branch, if it exists. If not, prompts for creation or
  # uses of current branch. If can't checkout because of an error, do something else?
  let flag = if create: "-b " else: ""
  withDir(repo):
    result = gitExecCmd("git checkout {flag}{branch}").isSome

# ...
```

tinygiant98 commented:
Okay, that sounds good.  I'll learn that syntax and get it going.

tinygiant98 commented:
Option changes are in.  Had to modify `empty` also to work properly since it returns a value even when it "fails".  No error is thrown because, technically, it didn't really fail, so had to account for the empty string.  Also, changed the error message to discount the git error message and point the user to their repo for resolution.  Is there a way to use `Option` to retrieve the git error message and be able to tell the difference between valid (passing) output and invalid (failed) error messages?  Didn't seem to be, but still learning this one.

squattingmonk commented:
If you want to get an error and it's associated output, use `execCmdEx` from `osproc`. It lets you get the error code and output.

tinygiant98 commented:
So the code should be good as-is with the `Option` if you want to keep it that way.  Another option I was looking at was to allow the passage of errors without making additional calls, essentially doing away with `gitExecCmd` and letting the individual procedures handle the error/display.  Something like this:
```nim
proc gitUser*: string =
  let gitResult = execCmdEx("git config --get user.name")

  if gitResult.exitCode == 0:
    gitResult.output.strip
  else: "whatever the default is"
```
That gives the procedure the option to display the error, if set up differently, to pass the error value back.  Probably best to leave it here, so here's what I came up with as an example:
```nim
proc checkout(branch: string, repo: string, create = false, throw = false): bool = 
  let flag = if create: "-b " else: ""
  withDir(repo):
    let gitResult = gitExecCmdS(fmt"git checkout {flag}{branch}")
    
    result = gitResult.exitCode == 0
    if not result and throw:
      error(gitResult.output.strip)
```
This stuff isn't absolutely necessary, just looking to maximize error handling and minimize additional code to allow better error handling for future expansion (such as adding hg, cloning repositories, etc. when we might need to directly report the vcs error to the user).  Seems running a git command to return the default, then running it again to grab the error could be streamlined.  I'm sure there's a better way to handle this, just need to learn it.  Let me know if you think this is important or if I'm heading down a rabbit hole (happens a lot when I'm learning something new).

tinygiant98 commented:
Sorry, didn't squash those commits.  Moved to function-level error handling to return defaults and still allow the display of errors when desired.  Also added some logic to `checkout` to prevent accidentally creating branches from branches other than master.

squattingmonk commented:
I like this. Much cleaner way to do it. I rebased onto master and removed a stray debug message.
